src/Kestrel.Transport.Sockets/Internal/SocketAwaitableEventArgs.cs

```diff
@@ -5,74 +5,137 @@
+        }
+
+        public void OnCompleted(Action<object> continuation, object state, short token, ValueTaskSourceOnCompletedFlags flags)
+        {
+            // We're ignoring ValueTaskSourceOnCompletedFlags and token since this is used in a single place in the code base
+            // so we don't need to handle capturing the SynchronizationContext and ExecutionContext
+            Debug.Assert((flags & ValueTaskSourceOnCompletedFlags.FlowExecutionContext) == 0, "FlowExecutionContext is set.");
+
+            // Use UserToken to carry the continuation state around
+            UserToken = state;
+            Action<object> awaitableState = Interlocked.CompareExchange(ref _completion, continuation, null);
 
-            if (SocketError != SocketError.Success)
+            if (ReferenceEquals(awaitableState, _callbackCompleted))
+            {
+#if NETCOREAPP2_1
+                ThreadPool.QueueUserWorkItem(continuation, state, preferLocal: false);
+#else
+                Task.Factory.StartNew(continuation, state);
+#endif
+            }
+            else if (awaitableState != null)
             {
-                ThrowSocketException(SocketError);
+                Debug.Fail("Multiple continuations registered!");
             }
 
-            return BytesTransferred;
+        }
+
+        public int GetResult(short token)
+        {
+            SocketError error = SocketError;
+            int bytes = BytesTransferred;
+
+            Reset();
+
+            if (error != SocketError.Success)
+            {
+                ThrowSocketException(error);
+            }
+
+            return bytes;
 
             void ThrowSocketException(SocketError e)
             {
                 throw new SocketException((int)e);
             }
         }
 
-        public void OnCompleted(Action continuation)
+        public ValueTask<int> ReceiveAsync(Socket socket)
         {
-            if (ReferenceEquals(_callback, _callbackCompleted) ||
-                ReferenceEquals(Interlocked.CompareExchange(ref _callback, continuation, null), _callbackCompleted))
+            if (socket.ReceiveAsync(this))
             {
-                Task.Run(continuation);
+                return new ValueTask<int>(this, 0);
             }
+
+            return CompleteSocketOperation();
         }
 
-        public void UnsafeOnCompleted(Action continuation)
+        /// <summary>Initiates a send operation on the associated socket.</summary>
+        /// <returns>This instance.</returns>
+        public ValueTask<int> SendAsync(Socket socket)
         {
-            OnCompleted(continuation);
+            if (socket.SendAsync(this))
+            {
+                return new ValueTask<int>(this, 0);
+            }
+
+            return CompleteSocketOperation();
         }
 
-        public void Complete()
+        private void Reset()
         {
-            OnCompleted(this);
+            Volatile.Write(ref _completion, null);
```

Comment:
Good catch. I guess theoretically without this volatile write, if `OnCompleted(SocketAsyncEventArgs)` was called from another thread before `OnCompleted(Action<object> continuation, ...`, the first OnCompleted call could observe a stale _completion callback and invoke it a second time.